app/services/arcgis_api/token_keeper.rb

```diff
@@ -0,0 +1,236 @@
+      # mixed usage patterns. Without this cache entries don't expire.
+      # More at https://api.rubyonrails.org/classes/ActiveSupport/Cache/RedisCacheStore.html
+      Rails.cache.try(:redis)&.expireat(cache_key, expires_at.to_i)
+    end
+
+    # Makes a request to retrieve a new token
+    # it expires after 1 hour
+    # @return [String] Auth token
+    def retrieve_token
+      token, expires = request_token.fetch_values('token', 'expires')
+      expires_at = Time.zone.at(expires / 1000).to_f
+      cache_entry = {
+        token: token,
+        expires_at: expires_at,
+      }
+
+      if IdentityConfig.store.arcgis_token_sliding_expiration_enabled == true
+        cache_entry = cache_entry.merge(
+          { sliding_expires_at: prefetch_ttl >= 0 ? expires_at - 3 * prefetch_ttl : expires_at },
+        )
+      end
+      save_token(cache_entry, expires_at)
+      cache_entry
+    end
+
+    def remove_token
+      Rails.cache.delete(cache_key)
+    end
+
+    private
+
+    def handle_expired_token(cache_value)
+      cache_value = wrap_raw_token(cache_value) unless cache_value.is_a?(Hash)
+      expires_at = cache_value.fetch(:expires_at, nil)
+      sliding_expires_at = cache_value&.fetch(:sliding_expires_at, nil)
+      # start fetch new token before actual expires time
+      if IdentityConfig.store.arcgis_token_sliding_expiration_enabled
+        now = Time.zone.now.to_f
+        if sliding_expires_at && now >= sliding_expires_at
+          # if passed sliding expiration time(exp-3*prefetch_ttl)
+          # but not 2*prefetch_ttl from hard expiration time
+          if now > (sliding_expires_at + prefetch_ttl)
+            cache_value = retrieve_token
+          else
+            # extend sliding expiration time
+            cache_value[:sliding_expires_at] = sliding_expires_at + prefetch_ttl
+            save_token(cache_value, expires_at)
+          end
+        end
+      elsif expires_at && expires_at <= now
+        cache_value = retrieve_token
+      end
+      cache_value
+    end
+
+    # Makes HTTP request to authentication endpoint and
+    # returns the token and when it expires (1 hour).
+    # @return [Hash] API response
+    def request_token
+      body = {
+        username: IdentityConfig.store.arcgis_api_username,
+        password: IdentityConfig.store.arcgis_api_password,
+        referer: IdentityConfig.store.domain_name,
+        f: 'json',
+      }
+
+      connection.post(
+        IdentityConfig.store.arcgis_api_generate_token_url, URI.encode_www_form(body)
+      ) do |req|
+        req.options.context = { service_name: 'arcgis_token' }
+      end.body
+    end
+
+    def connection
+      faraday_retry_options = {
+        retry_statuses: (400..500).to_a,
+        max: IdentityConfig.store.arcgis_get_token_retry_max,
+        methods: %i[post],
+        interval: IdentityConfig.store.arcgis_get_token_retry_interval,
+        backoff_factor: IdentityConfig.store.arcgis_get_token_retry_backoff_factor,
```

Comment:
I recommend also setting `interval_randomness` here, esp. since we don't have concurrency controls enabled for this job.

Reply:
corrected.